src/compiler/program.ts

```diff
@@ -904,7 +904,13 @@ namespace ts {
             let declarationDiagnostics: Diagnostic[] = [];
 
             if (options.noEmit) {
-                return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emittedFiles: undefined, emitSkipped: true };
+                // Allow only emitting of declarations
+                if (options.declaration) {

```

Comment:
Hm, this is interesting. The user wrote `noEmit`, but you're still emitting `.d.ts` files. Is there a better mechanism to surface this behavior? @mhegazy @bowdenk7